### PR TITLE
Use proc macro to define Config

### PR DESCRIPTION
**Pull request by topecongiro**
This PR implements a proc macro for defining `Config` struct, which replaces the current `define_config` macro.

The ultimate goal is to define the config like the following:

```rust
#[rustfmt_config]
pub struct Config {
    /// A maximum width of each line.
    ///
    /// - Default value: 100
    /// - Possible values: any positive integers
    /// - Stable: Yes
    ///
    /// See also err_on_line_overflow.
    ///
    /// ## Example
    ///
    /// ### Input
    ///
    /// ```rust
    /// // An example code.
    /// ```
    ///
    /// ### 100 (default)
    /// ```rust
    /// // An example code formatted with max_width = 100.
    /// ```
    ///
    /// ### 80 (default)
    /// ```rust
    /// // An example code formatted with max_width = 80.
    /// ```
    #[rustfmt::config(default = 100, stable = "1.0.0")]
    max_width: usize,

    // ... other configs
}
```

and let the config proc macro generate Configurations.md and test code under test/source/configs/ and tests/target/configs/. This PR is the first step toward the goal.

Other minor benefits include the following:

- a better interface for setters
- internal config options are now private

Since the `Config` is part of the public API of rustfmt-as-a-lib, and this PR changes its interface, we should merge this as a part of the preparation for 2.0 release. I am creating a PR at the moment, however, for early feedback.

### CODE REVIEW

File: config_proc_macro/src/define_config.rs [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+            UnstableConfigOption(Vec<&'static str>),
+            TomlParseError(toml::de::Error),
+            TomlSerializationError(toml::ser::Error),
+            InvalidKeyValuePair(String, String),
+        }
+
+        impl std::fmt::Display for ConfigError {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+                match self {
+                    ConfigError::UnstableConfigOption(unstables) => {
```
Comment (calebcartwright) [inline config_proc_macro/src/define_config.rs]: This will resolve #3872 correct?

File: src/bin/main.rs [hunk 2]
```diff
@@ -243,11 +242,11 @@ fn format_string(input: String, options: GetOptsOptions) -> Result<i32, FailureE
     let (mut config, _) = load_config(Some(Path::new(".")), Some(options.clone()))?;
 
     // emit mode is always Stdout for Stdin.
-    config.set().emit_mode(EmitMode::Stdout);
-    config.set().verbose(Verbosity::Quiet);
+    config.set_emit_mode(EmitMode::Stdout);
```
Comment (calebcartwright) [inline src/bin/main.rs]: 👍 definitely a friendly interface on getters/setters

File: src/config/mod.rs [hunk 3]
```diff
@@ -24,155 +27,359 @@ pub(crate) mod file_lines;
...
-         via the --file-lines option";
-    width_heuristics: WidthHeuristics, WidthHeuristics::scaled(100), false,
-        "'small' heuristic values";
-    emit_mode: EmitMode, EmitMode::Files, false,
-        "What emit Mode to use when none is supplied";
-    make_backup: bool, false, false, "Backup changed files";
-    print_misformatted_file_names: bool, false, true,
-        "Prints the names of mismatched files that were formatted. Prints the names of \
-         files that would be formated when used with `--check` mode. ";
+    /// How much to information to emit to the user.
```
Comment (calebcartwright) [inline src/config/mod.rs]:
```suggestion
    /// How much information to emit to the user.
```

File: config_proc_macro/src/define_config.rs (hunk 1 above)
Comment (topecongiro) [reply config_proc_macro/src/define_config.rs]: Unfortunately no, IIRC the issue is that we are parsing the same rustfmt.toml and emitting errors multiple times.

**Comment by topecongiro** [comment]
I will try this again after 2.0 release :man_shrugging:

**Pull request closed without merging**